Added the ability to add/retrieve local resources in MapReduce and Spark Programs

bdmogal opened the pull request:
Exposed via 
- ``ClientLocalizationContext`` - Allows to localize resources. Available in ``MapReduceContext`` and ``SparkContext``.
- ``TaskLocalizationContext`` - Allows to retrieve localized resources. Available in ``MapReduceTaskContext`` and ``ExecutorSparkContext``.

Jira: [CDAP-4135](https://issues.cask.co/browse/CDAP-4135)
Build: http://builds.cask.co/browse/CDAP-DUT3030

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/LocalizationContext.java:
```diff
@@ -0,0 +1,42 @@
+   *
+   * @param uri URI of the file to localize
+   */
+  void addLocalFile(URI uri);
```
Should it be just `localize`? Also, you might want to take a name and a URI. Also, you'll also have variants for user to specific if it is an archive or not so that the distributed cache will do the unarchive.

bdmogal replied:
Sure, will do. So ``localize(String name, URI uri)`` will delegate to ``localize(URI.create(String.format("%s#%s", uri, name)))``?

I'll add the archive methods. They will just delegate to ``job.addCacheArchive()`` for MapReduce, and set ``LocalizeResource.archive`` to true for Spark.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceLifecycleContext.java:
```diff
@@ -126,6 +127,16 @@ public String getNamespace() {
 
   @Override
+  public void addLocalFile(URI uri) {
+    throw new UnsupportedOperationException("Local cache files are not supported in Mapper/Reducer tasks.");
```
Why not just call `delegate` like other methods do?

bdmogal replied:
This is used in a Map/Reduce task, right? I thought we should only allow retrieving, and not adding in tasks. Adding should only be allowed in job setup/beforeSubmit?

chtyim replied:
But the contract of this class is to delegate. The context that this delegate to will do the right thing.

bdmogal replied:
I see what you mean. Yes, this should just delegate, and ``BasicMapReduceTaskContext`` should handle it. Will fix.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/AbstractSparkContext.java:
```diff
@@ -83,7 +86,7 @@ protected AbstractSparkContext(ApplicationSpecification applicationSpecification
                                  MetricsContext metricsContext, LoggingContext loggingContext,
                                  @Nullable PluginInstantiator pluginInstantiator,
-                                 @Nullable WorkflowToken workflowToken) {
+                                 @Nullable WorkflowToken workflowToken, @Nullable List<URI> localResources) {
```
Why it comes from the constructor?

chtyim replied:
And why nullable?

chtyim replied:
Also, since only `ClientSparkContext` supports localization, the `AbstractSparkContext` shouldn't have anything to do with it.

bdmogal replied:
Even in the ``ExecutionSparkContext``, we should be able to do ``getLocalFiles``, right?

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/LocalizationContext.java:
```diff
@@ -0,0 +1,42 @@
+  /**
+   * @return a list of files localized for this program
+   */
+  List<URI> getLocalFiles() throws IOException;
```
How is it being used?

chtyim replied:
For this interface to be more useful, I think it should look something like this:

```java
interface LocalizationContext {
  void localize(String name, URI uri);
  void localize(String name, URI uri, boolean archive);

  File getLocalizedFile(String name);
  Map<String, File> getAllLocalizedFiles();
}
```

chtyim replied:
And given that the context we do set vs context we do get are different, it might make sense to split this interface into two.

bdmogal replied:
It is used to retrieve local files in map-reduce/spark programs. So the way I implemented it, even to retrieve local files, users will have to go through CDAP APIs, as opposed to doing ``job.getLocalCacheFiles`` (I am not sure of the Spark equivalent)

bdmogal replied:
See https://github.com/caskdata/cdap/pull/4414/files#diff-7718c4ed3373be555855f1f0acae0db3R97

bdmogal replied:
I see. Does this seem ok?
``` 
interface TaskLocalizationContext {
  File getLocalizedFile(String name);
  Map<String, File> getAllLocalizedFiles();
}

interface ClientLocalizationContext extends TaskLocalizationContext {
  void localize(URI uri);
  void localize(URI, boolean archive);
  void localize(String name, URI uri);
  void localize(String name, URI uri, boolean archive);
}
```

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceTaskContext.java:
```diff
@@ -35,7 +36,7 @@
 @Beta
 public interface MapReduceTaskContext<KEYOUT, VALUEOUT> extends RuntimeContext, DatasetContext,
-  ServiceDiscoverer, PluginContext {
+  ServiceDiscoverer, PluginContext, LocalizationContext {
```
Why the task context need to support localization? And how? Task context is only used in Mapper/Reducer, which cannot localize file anymore.

bdmogal replied:
Yes, it shouldn't support localizing new files, but we should be able to retrieve localized files in the task context, right?

chtyim replied:
fair enough.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/AppWithLocalFiles.java:
```diff
@@ -0,0 +1,137 @@
+          String localFilePath = args.get(STOPWORDS_FILE_ARG);
+          for (URI localFile : localFiles) {
+            if (localFilePath.equals(localFile.toString())) {
+              try (FileSystem fs = FileSystem.get(localFile, new Configuration());
```
Since it's a local file, you don't need to go through `FileSystem`. Just use `new File(localFile)`.

chtyim replied:
Better yet, `Files.newBufferedReader(new File(localFile).toPath(), Charsets.UTF_8)`.

chtyim replied:
`java.nio.file.Files`

bdmogal replied:
So, I used it because I'm passing the I couldn't get ``java.io.File`` to find a file of the format ``file:/path/to/file``, which is what I get here from https://github.com/caskdata/cdap/pull/4414/files#diff-0642c8ed9095b2017093f641afd4fc7bR572

chtyim replied:
See my comment about the interface. We shouldn't be returning the original URI used for localization, right? It should always refer to a file local to the container.

bdmogal commented:
@chtyim have re-vamped this PR per our discussion. Can you please take another look?

bdmogal commented:
Tested on cluster https://github.com/bdmogal/cdap-applications/blob/master/src/main/java/my/bhooshan/cdap/WordCountWithStopwords.java

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/ClientLocalizationContext.java:
```diff
@@ -0,0 +1,67 @@
+   *
+   * @param uri the {@link URI} of the file to localize
+   */
+  void localize(URI uri) throws URISyntaxException;
```
I think we should always require a name. Meaning this and the `localize(URI uri, boolean archive)` are not needed.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/ClientLocalizationContext.java:
```diff
@@ -0,0 +1,67 @@
+   *             runtime using {@link TaskLocalizationContext#getLocalFile(String)}
+   * @param uri the {@link URI} of the file to localize
+   */
+  void localize(String name, URI uri) throws URISyntaxException;
```
Why it throws a `URISyntaxException`? The user is already passing in an `URI` object, it can't be having URISyntaxException, right?

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/TaskLocalizationContext.java:
```diff
@@ -0,0 +1,56 @@
+   * @param archive indicates if the requested file is an archive
+   * @return a {@link File} for the localized file
+   */
+  File getLocalFile(String name, boolean archive) throws IOException;
```
It shouldn't master if it is archive or not. The `File` returned by the `getLocalFile` would be a directory if it is an archive.

bdmogal replied:
But for example, we need to decide whether to call ``Job.getLocalCacheFiles()`` or ``Job.getLocalCacheArchives`` in the case of map-reduce, right? Or are you saying always call ``Job.getLocalCacheFiles()``?

chtyim replied:
We should preserve those information ourselves, not requiring the user to provide that.

chtyim replied:
We shouldn't design API based on implementation.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/TaskLocalizationContext.java:
```diff
@@ -0,0 +1,56 @@
+   * @param name the local file's name
+   * @return a {@link File} for the localized file
+   */
+  File getLocalFile(String name) throws IOException;
```
Instead of `IOException`, should be just `FileNotFoundException`, since this method shouldn't be doing any IO stuff.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/TaskLocalizationContext.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.Map;
+
+/**
+ * Localization context for Program Runtimes (MapReduce Tasks/Spark Executors).
```
Better not to mention MapReduce/Spark in the interface.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/TaskLocalizationContext.java:
```diff
@@ -0,0 +1,41 @@
+public interface TaskLocalizationContext {
+
+  /**
+   * @param name the local file's name
```
Javadoc for method without any description looks awkward in the javadoc html. Add some description. E.g. `Returns a {@link File} representing the path to a localized file localized with the given name`.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/spark/SparkContext.java:
```diff
@@ -211,4 +213,9 @@
   WorkflowToken getWorkflowToken();
+
+  /**
+   * @return the {@link TaskLocalizationContext} for the {@link Spark} program
```
Need to document that the returned instance can be used inside Closure and that it's Serializable.

chtyim replied:
See the doc of getter of ServiceDiscover/Metrics/PluginContext.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/DefaultTaskLocalizationContext.java:
```diff
@@ -0,0 +1,50 @@
+  public File getLocalFile(String name) throws FileNotFoundException {
+    if (!localizedResources.containsKey(name)) {
+      throw new FileNotFoundException(String.format("The specified file %s was not found. Please make sure it was " +
+                                                      "localized using context.localize in the Spark program's " +
```
This class is only for Spark?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/DefaultTaskLocalizationContext.java:
```diff
@@ -0,0 +1,50 @@
+  private final Map<String, File> localizedResources;
+
+  public DefaultTaskLocalizationContext(Map<String, File> localizedResources) {
+    this.localizedResources = localizedResources;
```
Since the map will be returned through the `getAllLocalFiles` method, it's better to do a defensive immutable map copy.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceTaskContext.java:
```diff
@@ -107,6 +111,7 @@ public BasicMapReduceTaskContext(Program program,
     this.loggingContext = createLoggingContext(program.getId(), runId);
     this.spec = spec;
     this.plugins = Maps.newHashMap(program.getApplicationSpecification().getPlugins());
+    this.localizedResources = localizedResources;
```
Better construct a `DefaultTaskLocalizationContext` and delegate to that so that all program types share the same underlying implementation.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java:
```diff
@@ -255,6 +260,14 @@ private void setInputSelection(@Nullable List<Split> splits) {
+  }
+
+  public Map<String, File> getLocalizedResources() {
+    return GSON.fromJson(hConf.get(HCONF_ATTR_LOCAL_FILES), new TypeToken<Map<String, File>>() { }.getType());
```
Why serialize the `File`? The path won't be valid in the task container.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -1005,4 +1013,57 @@ private ClassLoader setContextCombinedClassLoader(BasicMapReduceContext context)
+        File tempFile = new File(targetDir, name);
+        localizedFile = new File(uri.getPath());
+        LOG.debug("Copy file from {} to {}", uri, tempFile);
+        Files.copy(new InputSupplier<InputStream>() {
```
Use `Files.copy(File from, File to)`.

chtyim replied:
Also, if the resource is a archive, shouldn't you expand the uncompress the file (unjar, ungzip, un-tgz)?

bdmogal replied:
I will add the decompression in a follow up PR - https://issues.cask.co/browse/CDAP-4178

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -1005,4 +1013,57 @@ private ClassLoader setContextCombinedClassLoader(BasicMapReduceContext context)
+    Map<String, LocalizeResource> resourcesToLocalize = context.getResourcesToLocalize();
+    for (final Map.Entry<String, LocalizeResource> entry : resourcesToLocalize.entrySet()) {
+      final File localizedFile;
+      LocalizeResource resourceToLocalize = entry.getValue();
```
This naming is very confusing. It has a `resourcesToLocalize` as a Map outside the loop. In fact, this variable only used once, so better chain the call:

```java
URI uri = entry.getValue().getURI();
```

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -1005,4 +1013,57 @@ private ClassLoader setContextCombinedClassLoader(BasicMapReduceContext context)
+      }
+      localizedResources.put(name, localizedFile);
+      // also serialize the local resource names, so they can be recreated later.
+      mapredConf.set(Constants.Localization.LOCAL_RESOURCES, Joiner.on(",").join(resourcesToLocalize
```
Why needed? I think you already serialize the map through the `MapReduceContextConfig` class?

bdmogal replied:
not needed. forgot to clean this up. removed.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java:
```diff
@@ -499,4 +507,18 @@ private Configuration configureStreamInput(Configuration configuration,
     return dataset;
   }
+
+  public TaskLocalizationContext getTaskLocalizationContext() {
```
`@Override`?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkRuntimeService.java:
```diff
@@ -405,6 +416,47 @@ private File saveHConf(Configuration hConf, File targetDir) throws IOException {
+      final File localizedFile;
+      if (contextConfig.isLocal()) {
+        // in local mode, also localize resources in a temporary directory
+        File tempFile = new File(targetDir, entry.getKey());
```
You should refactor this local file copying mechanism to be shared with MapReduce as well.

bdmogal commented:
@chtyim addressed all comments but https://github.com/caskdata/cdap/pull/4414#discussion_r43798062. Will fix that in a follow up PR right after this one.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java:
```diff
@@ -1005,4 +1010,47 @@ private ClassLoader setContextCombinedClassLoader(BasicMapReduceContext context)
+    Map<String, LocalizeResource> resourcesToLocalize = context.getResourcesToLocalize();
+    for (Map.Entry<String, LocalizeResource> entry : resourcesToLocalize.entrySet()) {
+      String localizedFilePath;
+      URI uri = entry.getValue().getURI();
```
It's only use in else. Move it there.

bdmogal replied:
Fixed.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkRuntimeService.java:
```diff
@@ -139,18 +140,23 @@ protected void startUp() throws Exception {
           hConf.set(Constants.Plugin.ARCHIVE, pluginJar.getName());
         }
+        // also serialize the local resource names, so they can be recreated later.
+        hConf.set(Constants.Localization.LOCAL_RESOURCES, new Gson().toJson(localizedFiles.keySet()));
```
For constants only use for configuration exchange between SparkRuntimeService and SparkContextProvider, they are defined in SparkContextProvider, not in Constants.

bdmogal replied:
Fixed. Also made the constants in SparkContextProvider package-private since they are not required outside the spark package.

chtyim commented:
Just one comment left. rest LGTM.

chtyim commented:
LGTM. Please squash before merge.

bdmogal commented:
Squashed commits, build is green. Verified on cluster again. Merging.
